diff --git a/llvm/lib/Target/RISCV/RISCVInstrInfoXpulp.td b/llvm/lib/Target/RISCV/RISCVInstrInfoXpulp.td
index 090598a2037e..c3f5abf4db39 100644
--- a/llvm/lib/Target/RISCV/RISCVInstrInfoXpulp.td
+++ b/llvm/lib/Target/RISCV/RISCVInstrInfoXpulp.td
@@ -1589,6 +1589,9 @@ def : Pat<(v4i8 (build_vector GPR:$rs1, GPR:$rs2, GPR:$rs3, GPR:$rs4)),
 def : Pat<(v2i16 (splat_vector GPR:$rs1)), (PV_PACK_H GPR:$rs1, GPR:$rs1)>;
 def : Pat<(v4i8 (splat_vector GPR:$rs1)), (PV_ADD_SC_B X0, GPR:$rs1)>;
 
+def : Pat<(v4i8 (riscv_vmv_v_x_vl (v4i8 undef), GPR:$imm, VLOpFrag)),
+	   (PV_ADD_SC_B X0, (ADD (SLL X0, $vl), $imm))>;
+
 defm : GeneralSVectorPattern<seteq, "CMPEQ">;
 defm : GeneralSVectorPattern<setne, "CMPNE">;
 defm : GeneralSVectorPattern<setgt, "CMPGT">;
